sanic/reloader_helpers.py

```diff
@@ -85,18 +100,22 @@ def interrupt_self(*args):
         while True:
             need_reload = False
 
-            for filename in _iter_module_files():
-                try:
-                    mtime = os.stat(filename).st_mtime
-                except OSError:
-                    continue
-
-                old_time = mtimes.get(filename)
-                if old_time is None:
-                    mtimes[filename] = mtime
-                elif mtime > old_time:
-                    mtimes[filename] = mtime
-                    need_reload = True
+            for collection in (_iter_module_files(), *app.reload_dirs):
+                if isinstance(collection, Path):
+                    collection = collection.glob("**/*")
+
+                for filename in collection:
+                    try:
+                        check = _check_file(filename, mtimes)
+                    except OSError:
+                        continue
+
+                    if check:
+                        need_reload = True
+                        break
```

Comment:
Shouldn't this be inside the try block?

Also the two level iteration is a bit convoluted - I'm sure you can find a cleaner way to do it. Flatten the iterators, itertools.chain or something?

Reply:
IIRC, one needs to iterate over all the files, updating timestamps, and not return early even when differences are found. This is to avoid multiple reloads when many files are changed, so please pay attention to preserving that feature (fixed in 2019 probably).

I did not properly review this yet but overall it seems OK, especially if you double check the reloader loops and the code within. Possibly no changes required, I just cannot wrap my head around it right now.

Reply:
> Flatten the iterators, itertools.chain or something?

Yeah... that also would give us the ability to add files individually. I was trying to sneak this change in without being too invasive and then leave another opportunity to overhaul how the whole module works with better iteration.

> Shouldn't this be inside the try block?

Will check :eyes:

> IIRC, one needs to iterate over all the files, updating timestamps, and not return early even when differences are found. This is to avoid multiple reloads when many files are changed, so please pay attention to preserving that feature (fixed in 2019 probably).

Good catch :+1:

Reply:
> Shouldn't this be inside the try block?

You mean because of scoping? Variables are not scoped inside `try`

```
>>> try:
...  foo = 1
... except Exception:
...  ...
... 
>>> print(foo)
1
```

Reply:
I'm not sure if this functionally is any different. The point is we still need multiple iterators since we need to generate new globs on every run.

```python
            for filename in itertools.chain(
                _iter_module_files(),
                *(d.glob("**/*") for d in app.reload_dirs),
            ):
```

Reply:
> > Shouldn't this be inside the try block?
> You mean because of scoping? Variables are not scoped inside `try`

Nevermind. The `continue` in the exception handler takes care that this code is not reached when `check` is not set, which was my concern.